File: Python3/797. All Paths From Source to Target.py
class Solution:
    def allPathsSourceTarget(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: List[List[int]]
        """
        n = len(graph)
        if len(graph) == 1:
            return [[0]]
        ans = []

        def aPST(path, node):
            """
            :type path: list[int]
            :type node: int
            """
            if node == n - 1:
                ans.append(path)
                return
            for i in graph[node]:
                aPST(path + [i], i)

        aPST([0], 0)
        return ans


sol = Solution()
print(sol.allPathsSourceTarget([[1, 2], [3], [3], []]))
print(sol.allPathsSourceTarget([[3], [0, 1, 3], [], [2]]))
print(sol.allPathsSourceTarget([[3], [0, 1, 3], [], [2]]))
print(sol.allPathsSourceTarget([[1, 2, 3], [2], [3], []]))

File: 797. All Paths From Source to Target.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        vector<vector<int>> ans;
        vector<int> path;
        dfs(graph, 0, path, ans);
        return ans;
    }
    
private:
    void dfs(vector<vector<int>>& graph, int currNode, vector<int>& path, vector<vector<int>>& ans) {
        if (currNode == graph.size() - 1) {
            path.push_back(currNode);
            ans.push_back(path);
            path.pop_back();
            return;
        }
        path.push_back(currNode);
        for (int node: graph[currNode]) {
            dfs(graph, node, path, ans);
        }
        path.pop_back();
    }
};

int main() {
    vector<vector<int>> graph{{1,2}, {3}, {3}, {}};
    auto ans = Solution().allPathsSourceTarget(graph);
    for (auto& v: ans) {
        for (int i: v) cout << i << ' ';
        cout << endl;
    }
    return 0;
}
